Refactor Hilbert_basis() and replace a slow test

orlitzky opened the pull request:
Initially my goal was to replace one slow `Hilbert_basis()` test, but I did some refactoring along the way -- none of which really affects the performance of the Hilbert basis calculation. The commit message lists these changes.

To replace the test, I made up some examples and fed them to normaliz. If the sage answer agrees with the normaliz answer, they must both be right, right?

orlitzky committed:
src/sage/geometry/cone.py: refactor Hilbert_basis()

Some refactorings to the implementation of Hilbert_basis():

  * Construct a cone L from our linear_subspace() so that "y in L"
    works as intended (currently we try to coerce y into a vector
    space in a try/except block). This is not any faster, but it
    makes the code easier to read.

  * Remove the irreducibles from "gens" as we construct it.

  * Negate a condition in a loop to avoid bailing out with "continue"
    as part of the normal control flow.

  * Use a boolean indicator to check if the list of irreducibles
    was modified, rather than recomputing its length.

orlitzky committed:
src/sage/geometry/cone.py: faster Hilbert_basis() tests

We have one Hilbert_basis() test that is raising "slow test!" warnings
at around 40s. Here we replace it with three tests, each of which runs
relatively quickly. The trio completes in about 15s.

Since I know very little about Hilbert bases, I have checked the
results using Normaliz. For example,

  $ cat cone.in
  amb_space 4
  cone 4
  1 0 1 0
  -1 0 1 0
  0 1 1 0
  0 -1 1 0
  $ normaliz --HilbertBasis cone
  $ cat cone.out
  ...

(the resulting basis is written to cone.out).

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -2274,10 +2274,11 @@ def dual(self):
+            rays = chain(self.facet_normals(),
+                         ( c*r
+                           for r in self.orthogonal_sublattice().gens()
+                           for c in (1,-1) ))
```
I don't see how this could affect performance: `Cone` with `check=True` first feeds its first argument to `normalize_rays`, which in turn applies `list` as first thing.

orlitzky replied:
It doubt it has any noticeable impact. I was simply poking around and noticed that this list is created, appended to multiple times, and then immediately discarded:

```
rays = list(self.facet_normals())
for ray in self.orthogonal_sublattice().gens():
    rays.append(ray)
    rays.append(-ray)
self._dual = Cone(rays, ...
```

Using `itertools.chain()` instead allows us to iterate though both "lists" individually, one after the other, without having to combine them both into one big list first. The difference would probably only be noticeable if the lists were absolutely huge.

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -4436,40 +4450,39 @@ def Hilbert_basis(self):
+            # Our linear_subspace(), but as a cone, so that
+            # containment testing using "in" works properly.
+            L = Cone( (c*r for c in (1,-1) for r in self.lines()),
+                      self.lattice() )
```
```suggestion
            L = Cone((c*r for c in (1, -1) for r in self.lines()),
                     self.lattice())
```

orlitzky replied:
done

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -4436,40 +4450,39 @@ def Hilbert_basis(self):
-                pass
+        irr_modified = False  # have we appended to "irreducible"?
+        gens = [ x for x in self.semigroup_generators()
+                 if x not in irreducible ]
```
```suggestion
        gens = [x for x in self.semigroup_generators()
                if x not in irreducible]
```

orlitzky replied:
done

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -4436,40 +4450,39 @@ def Hilbert_basis(self):
-            return self.rays()
-        else:
+            if all( (y in L) or (x-y not in self)
+                    for y in chain(irreducible,gens) ):
```
```suggestion
            if all((y in L) or (x-y not in self)
                   for y in chain(irreducible, gens)):
```

orlitzky replied:
done

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -4436,40 +4450,39 @@ def Hilbert_basis(self):
+            # Avoid the PointCollection overhead if nothing was
+            # added to the irreducible list beyond self.rays().
+            return self.rays()
+
```
```suggestion
        # Avoid the PointCollection overhead if nothing was
        # added to the irreducible list beyond self.rays().
        return self.rays()
```

orlitzky replied:
Good idea, thanks. I also made a similar change at the beginning of the function. Defining `L = ()` unconditionally does not take any time and lets us eliminate one branch of the if/else.

orlitzky committed:
src/sage/geometry/cone.py: reviewer suggestions in Hilbert_basis()

Tighten the whitespace around list/generator comprehensions, and
simplify the control flow in two instances by eliminating an if/else
branch. (Only one of these was suggested by the reviewer, but the
other is in a similar spirit.) Thanks to Martin Rubey for the
suggestions.

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -2274,10 +2277,11 @@ def dual(self):
+            rays = chain(self.facet_normals(),
+                         ( c*r
+                           for r in self.orthogonal_sublattice().gens()
+                           for c in (1,-1) ))
```
I suggest to revert this change if it does not affect performance.  I find it less readable.  Alternatively, you could replace the two appends with `rays.extend((ray, -ray))`.

If you decide to keep the change, the whitespace should be corrected.

orlitzky replied:
Ok, I did a force push to drop this commit.

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -4436,41 +4453,41 @@ def Hilbert_basis(self):
-            return self.rays()
-        else:
+            if all((y in L) or (x-y not in self)
+                    for y in chain(irreducible, gens)):
```
```suggestion
                   for y in chain(irreducible, gens)):
```

orlitzky replied:
And done.

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -4436,41 +4453,41 @@ def Hilbert_basis(self):
+        # Avoid the PointCollection overhead if nothing was
+        # added to the irreducible list beyond self.rays().
+        return self.rays()
+
```
```suggestion
```

orlitzky replied:
Done, thanks.

mantepse commented on src/sage/geometry/cone.py:
```diff
@@ -1686,11 +1686,14 @@ def _contains(self, point, region='whole cone'):
-            elif pr < 0 or need_strict and pr == 0:
+            if pr < 0:
                 return False
+            elif pr > 0:
```
```suggestion
            if pr > 0:
```

orlitzky replied:
I have dropped this commit as well for now. It turns out that the new version can be slower in some pathological cases (like the empty cone in a big space) where most of the constraints are equality and the equality constraints are listed first. It might be possible to work around, but I have other things I should be doing instead of trying to shave nanoseconds off of this method :)

orlitzky committed:
src/sage/geometry/cone.py: use Cone(check=False) in Hilbert_basis()

When constructing the subcone that represents the given cone's
linear_subspace(), we don't need to check that the generators are
valid or minimal -- fewer generators might work, but no subset will
work.

orlitzky committed:
src/sage/geometry/cone.py: faster containment for the trivial cone

When the cone K has no rays, the test "x in K" can be done quickly by
checking if x is zero. This can be a significant improvement if the
lattice is large, and risks wasting only as much time as it takes to
compare an integer to zero (i.e. nothing compared to how long the rest
of the containment test is going to take).
